src/libraries/drand/_index.md

```diff
@@ -0,0 +1,118 @@
+
+- cache the beacon's `Period`                           -- the period of time between each drand randomness generation
+- cache the beacon's `GenesisTime`                      -- at which the first round in the drand randomness chain is created
+- verify that the beacon's `PublicKey` is appropriate   -- ie that the filecoin node connected to the right drand beacon
+
+Thereafter, the Filecoin client can call drand's endpoints:
+
+- `LastPublic` to get the latest randomness value produced by the beacon
+- `Public` to get the randoomness value produced by the beacon at a given index
+
+{{<label drand>}}
+### Using drand in Filecoin
+
+Drand is used as a randomness beacon for leader election in Filecoin. You can read more about that in {{<sref leader_election>}}. See drand used in the Filecoin lotus implementation [here](https://github.com/filecoin-project/lotus/blob/master/chain/beacon/drand/drand.go).
+
+While drand returns multiple values with every call to the beacon (see above), Filecoin blocks need only store a subset of these in order to track a full drand chain. This information can then be mixed with on-chain data for use in Filecoin. See {{<sref randomness>}} for more.
+
+#### Verifying an incoming drand value
+
+Upon receiving a new drand randomness value from a beacon, a Filecoin node should immediately verify its validity. That is, it should verify:
+
+- that the `Signature` field is verified by the beacon's `PublicKey` as the beacon's signature of `SHA256(PreviousSignature || Round)`.
+- that the `Randomness` field is `SHA256(Signature)`.
+
+See [drand](https://github.com/drand/drand/blob/master/beacon/beacon.go#L63) for an example.
+
+#### Fetching the appropriate drand value while mining
+
+There is a deterministic mapping between a needed drand round number and a Filecoin epoch number.
+
+After initializing access to a drand beacon, a Filecoin node should have access to the following values:
+
+- `filEpochDuration`    -- the Filecoin network's epoch duration (between any two leader elections)
+- `filGenesisTime`      -- the Filecoin genesis timestamp
+- `filEpoch`            -- the current Filecoin epoch
+- `drandGenesisTime`    -- drand's genesis timestamp
+- `drandPeriod`         -- drand's epoch duration (between any two randomness creations)
+
+Using the above, a Filecoin node can determine the appropriate drand round value to be used for use in {{<sref leader_election>}} in an epoch using both networks' reliance on real time as follows:
+
+```go
+MaxBeaconRoundForEpoch(filEpoch) {
+    // determine what the latest Filecoin timestamp was from the current epoch number
+    var latestTs
+    if filEpoch == 0 {
+        latestTs = filGenesisTime
+    } else {
+        latestTs = ((uint64(filEpoch) * filEpochDuration) + filGenesisTime) - filEpochDuration
+    }
+    // determine the drand round number corresponding to this timestamp
+    dround := (latestTs - drandGenesisTime) / uint64(drandPeriod)
```

Comment:
This is incorrect and it is off by 1. @nikkolasg will verify this and check this in the code (Audit issue 2)

Reply:
Below are the two functions that drand uses to determine the mapping round <-> timestamp
The main relevant lines is:
```go
fromGenesis := now - genesis
nextRound = uint64(math.Floor(float64(fromGenesis)/period.Seconds())) + 1
```
in short, drand is creating the round 1 at genesis time, like filecoin, so a +1 is needed in the line in the spec.
Full function:
```go
func CurrentRound(now int64, period time.Duration, genesis int64) uint64 {
	nextRound, _ := NextRound(now, period, genesis)
	if nextRound <= 1 {
		return nextRound
	}
	return nextRound - 1
}

// NextRound returns the next upcoming round and its UNIX time given the genesis
// time and the period.
// round at time genesis = round 1. Round 0 is fixed.
func NextRound(now int64, period time.Duration, genesis int64) (nextRound uint64, nextTime int64) {
	if now < genesis {
		return 1, genesis
	}
	fromGenesis := now - genesis
	// we take the time from genesis divided by the periods in seconds, that
	// gives us the number of periods since genesis. We add +1 since we want the
	// next round. We also add +1 because round 1 starts at genesis time.
	nextRound = uint64(math.Floor(float64(fromGenesis)/period.Seconds())) + 1
	nextTime = genesis + int64(nextRound*uint64(period.Seconds()))
	return nextRound + 1, nextTime
}

Reply:
@nikkolasg understood. That said, I don't understand in the above code why you add `+1` twice in `NextRound`. Ergh that 1-indexing is confusing. I would suggest renaming the `nextRound` variable to `currentRound`, which is what it is, since it only becomes the `nextRound` after you increment it again in the return statement.